### PR TITLE
feat: Adjust user SDK

### PR DESCRIPTION
**Pull request by sfc-gh-asawicki**
Prepare user SDK for the user resource/datasource rework (continuation of https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/2936):
- add missing user parameters
- fix typos
- add missing integration tests (default creation, creation with all properties and parameters, altering, rename, tags)
- change a few types (mostly identifiers to the DefaultX values)
- document not working parts
- add snowflake objects assertions (manually)

What's not a part of this PR:
- new type of assertions could be added (assertion for parameters list)
- additions in 8.26 version of Snowflake (the docs were added today, so it will be a separate change)
- setting/unsetting different policies

### CODE REVIEW

File: pkg/sdk/testint/users_integration_test.go [hunk 1]
```diff
@@ -4,90 +4,185 @@ import (
...
+	email := random.Email()
+	newValue := random.AlphaN(5)
+	warehouseId := testClientHelper().Ids.WarehouseId()
+	schemaId := testClientHelper().Ids.SchemaId()
+	var schemaIdObjectIdentifier sdk.ObjectIdentifier = schemaId
+	// does not have to exist
+	roleId := testClientHelper().Ids.RandomAccountObjectIdentifier()
+	key, hash := random.GenerateRSAPublicKey(t)
+	key2, hash2 := random.GenerateRSAPublicKey(t)
+	_ = hash
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/users_integration_test.go]: Done

**Comment by mogren** [comment]
Any chance that TYPE can be a parameter here?

https://docs.snowflake.com/en/sql-reference/sql/create-user#label-user-type-property-person

The issue we have is that I can't create SERVICE users, which is a great security feature.

File: pkg/sdk/testint/users_integration_test.go [hunk 2]
```diff
@@ -161,68 +257,690 @@ func TestInt_UserCreate(t *testing.T) {
...
+			RSAPublicKey:   sdk.String(key),
+			RSAPublicKeyFp: sdk.String(hash),
+		}}
+
+		err := client.Users.Create(ctx, id, createOpts)
+		require.ErrorContains(t, err, "invalid property 'RSA_PUBLIC_KEY_FP' for 'USER'")
+
+		createOpts = &sdk.CreateUserOptions{ObjectProperties: &sdk.UserObjectProperties{
+			RSAPublicKey2:   sdk.String(key),
+			RSAPublicKey2Fp: sdk.String(hash),
```
Comment (sfc-gh-jmichalak) [inline pkg/sdk/testint/users_integration_test.go:389]: Isn't FP read-only? I think it means a key fingerprint which is computed from the provided key.

File: pkg/sdk/testint/users_integration_test.go [hunk 3]
```diff
@@ -161,68 +257,690 @@ func TestInt_UserCreate(t *testing.T) {
...
+		err := client.Users.Alter(ctx, user.ID(), alterOptions)
+		require.NoError(t, err)
+		t.Cleanup(testClientHelper().User.DropUserFunc(t, newID))
+
+		result, err := client.Users.Describe(ctx, newID)
+		require.NoError(t, err)
+		assert.Equal(t, newID.Name(), result.Name.Value)
+	})
+
+	t.Run("alter: set and unset", func(t *testing.T) {
```
Comment (sfc-gh-jmichalak) [inline pkg/sdk/testint/users_integration_test.go:619]: Nit: `set and unset object properties`.

File: pkg/sdk/testint/users_integration_test.go [hunk 4]
```diff
@@ -161,68 +257,690 @@ func TestInt_UserCreate(t *testing.T) {
...
+			HasMinsToUnlock("14").
+			HasDaysToExpiryNotEmpty().
+			HasComment("some comment").
+			HasDisabled(true).
+			HasMustChangePassword(true).
+			HasSnowflakeLock(false).
+			HasDefaultWarehouse(warehouseId.Name()).
+			HasDefaultNamespaceId(schemaId).
+			HasDefaultRole(roleId.Name()).
+			HasDefaultSecondaryRoles(`["ALL"]`).
```
Comment (sfc-gh-jmichalak) [inline pkg/sdk/testint/users_integration_test.go:678]: Are we planning to add sdk-level function to parse this field?

File: pkg/sdk/testint/users_integration_test.go [hunk 5]
```diff
@@ -161,68 +257,690 @@ func TestInt_UserCreate(t *testing.T) {
...
+		err = client.Users.Alter(ctx, id, alterOpts)
+		require.NoError(t, err)
+
+		assertParametersSet(id)
+
+		// unset is split into two because:
+		// 1. this is how it's written in the docs https://docs.snowflake.com/en/sql-reference/sql/alter-user#syntax
+		// 2. current implementation of sdk.UserUnset makes distinction between user and session parameters,
+		// so adding a comma between them is not trivial in the current SQL builder implementation
+		alterOpts = &sdk.AlterUserOptions{
```
Comment (sfc-gh-jmichalak) [inline pkg/sdk/testint/users_integration_test.go:807]: I think such behavior should be validated and tested.

**Review by sfc-gh-jmichalak — CHANGES_REQUESTED** [review]

File: pkg/sdk/testint/users_integration_test.go (hunk 2 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/users_integration_test.go:389]: Not according to the syntax: https://docs.snowflake.com/en/sql-reference/sql/create-user#syntax

File: pkg/sdk/testint/users_integration_test.go (hunk 4 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/users_integration_test.go:678]: TBH I don't know yet. Currently, as I understand the docs, only ["ALL"] value is supported, therefore I think we can just leave it with a TODO when more options are supported.

File: pkg/sdk/testint/users_integration_test.go (hunk 5 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/users_integration_test.go:807]: Added a unit test to check unset validation (and set validation too)

File: pkg/sdk/testint/users_integration_test.go (hunk 3 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/users_integration_test.go:619]: Done

**Review by sfc-gh-jmichalak — DISMISSED** [review]

**Review by sfc-gh-jmichalak — APPROVED** [review]

**Review by sfc-gh-jmichalak — APPROVED** [review]

**Pull request merged**